test/e2e/upgrades/ingress.go

```diff
@@ -0,0 +1,107 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package upgrades
+
+import (
+	. "github.com/onsi/ginkgo"
+	"net/http"
+
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/kubernetes/test/e2e/framework"
+)
+
+// IngressUpgradeTest adapts the Ingress e2e for upgrade testing
+type IngressUpgradeTest struct {
+	gceController *framework.GCEIngressController
+	jig           *framework.IngressTestJig
+	httpClient    *http.Client
+	ip            string
+}
+
+func (t *IngressUpgradeTest) Setup(f *framework.Framework) {
+	// jig handles all Kubernetes testing logic
+	jig := framework.NewIngressTestJig(f.ClientSet)
+
+	// gceController handles all cloud testing logic
+	gceController := &framework.GCEIngressController{
+		Ns:     f.Namespace.Name,
```

Comment:
Use framework.CreateNamespace to get a unique namespace name. Due to some limitations in how ginkgo allows me coordinate between test runs, this is sharing the framework object with other upgrade tests.